WordPress/gutenberg
# [RNMobile] Prevent "Undo Level" after Setting Featured Image via Image Block

This PR is intended as a temporary workaround for https://github.com/WordPress/gutenberg/issues/32937

`gutenberg-mobile`: https://github.com/wordpress-mobile/gutenberg-mobile/pull/3682

## Description

As outlined in https://github.com/WordPress/gutenberg/issues/32937, the undo/redo functionality doesn't work as expected after setting a featured image via the image block.

After setting a featured image via the block and then tapping undo, it appears as though the action has successfully reverted, as the **Featured** banner is removed from the block. However, the image still remains as the post's featured image.

With this PR, the issue is worked around by preventing an "undo level" from being created after a featured image is set via the image block. Tapping **undo** or **redo** immediately after setting a featured image via the image block should no longer have any effect.

Note, this PR doesn't solve the original issue. In fact, it will introduce a new UX issue, since the undo/redo buttons won’t work as the user would expect i.e. an undo will revert changes they made just before setting a featured image, rather than the featured setting, and may be confusing. However, this inconvenience should be considered preferable to the current experience in production, where changes to the featured image via the image block are not saved, despite the UI indicating the undo is successful. 

The **Future Considerations** section of this PR can be referred to for further hopes/considerations for any future iterations.

## How has this been tested?

* In the app, navigate to _Home_ → _Blog Posts_ to create a new blog post, one which doesn't already have a featured image set.
* Add a new image block and upload an image to it.
* Tap the image block's cog/gear icon to access its settings and then tap the **Set as Featured** button at the bottom.
* Verify that you see a confirmation message about the newly set featured image and that the block now has a **Featured** banner overlaying it.
* Tap the **undo** button in the app's toolbar.
* With this PR, tapping undo should have _no effect_ on the post's featured image. Verify that tapping **undo** or **redo** _doesn't_ change the featured image by observing whether there are any changes to the **Featured** banner and verifying that the image remains the same via _Post Settings_.

## Screenshots 

The following screenshots show the before and after footer in the image settings, with the updated footer noting that the undo/redo buttons won't work as expected for users:

| Before  | After |
| ------------- | ------------- |
| <img src="https://user-images.githubusercontent.com/2998162/113282237-8fecfc00-92de-11eb-9618-3f84c10e1720.png" width="100%"> | <img src="https://user-images.githubusercontent.com/2998162/128494892-5a44daa7-a097-445d-8beb-d86a0c6671c0.png" width="100%"> |

## Types of changes

This PR introduces a bug fix (a non-breaking change that fixes an issue), with the following main changes to the code:

* [A call to `editEntityRecord` is currently triggered](https://github.com/WordPress/gutenberg/pull/33057/files#diff-d45b169f7a1824d1f08790b97680e012bb30be66ed5714994a3959ba23cdee94R96) whenever a request to change the featured image is received from the apps (i.e. when the `Set/Remove as Featured` button is tapped from the image block). That action creator [includes an `undoIgnore` parameter](https://developer.wordpress.org/block-editor/reference-guides/data/data-core/#editEntityRecord) that makes it possible to ignore an edit in the undo history. For our purposes, that parameter is set to true so that a new undo level *isn't* created after the featured image is updated via the app.
* Additional text has also been added to the bottom of the image block's settings. [The additional text](https://github.com/WordPress/gutenberg/pull/33057/files#diff-7126ae03000a022bcefa0d2a5fb03b6aca56ed4047e3a49d2131babf229cf6c1R568) reads as follows: `Changes to featured image will not be affected by the undo/redo buttons `

## Future Considerations

As a more permanent fix for the original problem, the team discussed adding side effects to the undo/redo buttons, however, this deviates a lot from existing patterns in Gutenberg and is something we decided against pursuing.

The ideal fix would involve completely refactoring the existing data flow so that Gutenberg’s store is the primary location that the apps read from. A refactor would involve the following non-trivial changes:

* Refactoring the _Post Settings_ flow will require changes in the Android and iOS codebases, with careful testing to account for all the unknowns. There’s also the possibility of this raising a wider architectural debate around the _Post Settings_ panel in the apps, potentially including redesigning the whole user flow.
* The logic around a confirmation dialog currently exists on the native side of each app ([here](https://github.com/wordpress-mobile/WordPress-Android/blob/develop/libs/editor/WordPressEditor/src/main/java/org/wordpress/android/editor/gutenberg/GutenbergEditorFragment.java#L820) for Android and [here](https://href.li/?https://github.com/wordpress-mobile/WordPress-iOS/pull/16427/files#diff-beacfffc3f8a87edf94895da980a81e356f1b5aa7998f70791bc7baec9a11731R695) for iOS). This will need to be refactored so that it lives on the JS side. Again, this would be a desirable endpoint that would actually aid in implementing [this enhancement](https://github.com/WordPress/gutenberg/issues/30410). It has also been suggested that the confirmation dialog could be removed completely, which would be an interesting idea to explore with the design team.

## Checklist:
